### ISSUE TITLE
Like the cake, EventID is a lie

### ISSUE THREAD

**Issue by Yoric** [matrix-org/synapse#9777]
In the codebase, we have a type `EventID`. As of this day, it does not accurately represent event IDs. If I understand correctly, we have three different formats for event IDs and only one of them matches `EventID`. Attempting to shoehorn an event ID in another format into `EventID` turns out to be the wrong path. In other words, this is highly misleading.

I believe that we should fix `EventID` to accept these three formats and provide methods to ensure that the format is appropriate for a given context. The alternative would be to get rid of `EventID` and replace it with `str`.

**Comment by richvdh** [comment]
Yeah, EventID is barely used in the codebase. It *is* used in a couple of places to either generate or validate a V1 event ID, though those usages could probably be inlined without loss of generality.

Or we could rename EventID to `V1EventID` or something.

I don't think there would be much value in a class which can hold all three formats of event id.

**Comment by Yoric** [comment]
> I don't think there would be much value in a class which can hold all three formats of event id.

I agree, having both `V1EventID`, `V2EventID` and `V3EventID` (or perhaps something with nicer names) would probably be better for type-checking.
